src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -73,5 +92,28 @@ private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource
                 }
             }
         }
+
+        /// <summary>
+        /// When <paramref name="size"/> is very large, in many cases there will only be one chunk and that chunk will
+        /// be much smaller than <paramref name="size"/>. Therefore, it is worthwhile to build that chunk incrementally
+        /// rather than pre-allocating a <paramref name="size"/>-length array.
+        /// </summary>
+        private static bool TryCreateInitialChunkForLargeSize<TSource>(IEnumerator<TSource> enumerator, int size, [NotNullWhen(returnValue: true)] out TSource[]? chunk)
+        {
+            LargeArrayBuilder<TSource> builder = new();
```

Comment:
What made you pick `LargeArrayBuilder` over `List<T>`? It's only being used in Linq's speed optimizations specifically for `Enumerable.ToArray()` optimizations.

Reply:
A list might even be preferable since it's reusable and we know the capacity upper bound ahead of time. In any case, please use benchmarks to determine the best approach.

Reply:
> What made you pick LargeArrayBuilder over List<T>? It's only being used in Linq's speed optimizations specifically for Enumerable.ToArray() optimizations.

This felt like essentially the same operation as `ToArray()` so I just pulled it from that code. This could definitely be `List<T>` or `ArrayBuilder<T>` instead. As you say, if we use this approach for *every* chunk and not just the first one then `List<T>` is definitely preferable because we can re-use it.